cmd/sts-handlers.go

```diff
@@ -356,3 +367,130 @@ func (sts *stsAPIHandlers) AssumeRoleWithWebIdentity(w http.ResponseWriter, r *h
 func (sts *stsAPIHandlers) AssumeRoleWithClientGrants(w http.ResponseWriter, r *http.Request) {
 	sts.AssumeRoleWithJWT(w, r)
 }
+
+// AssumeRoleWithKerberosIdentity - implements authentication of a
+// user with Kerberos.
+func (sts *stsAPIHandlers) AssumeRoleWithKerberosIdentity(w http.ResponseWriter, r *http.Request) {
+	ctx := newContext(r, w, "AssumeRoleWithKerberosIdentity")
+
+	// Parse the incoming form data.
+	if err := r.ParseForm(); err != nil {
+		logger.LogIf(ctx, err)
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	// if r.Form.Get("Policy") != "" {
+	// 	writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+	// 	return
+	// }
+
+	if r.Form.Get("Version") != stsAPIVersion {
+		logger.LogIf(ctx, fmt.Errorf("Invalid STS API version %s, expecting %s", r.Form.Get("Version"), stsAPIVersion))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSMissingParameter))
+		return
+	}
+
+	action := r.Form.Get("Action")
+	switch action {
+	case kerberosIdentity:
+	default:
+		logger.LogIf(ctx, fmt.Errorf("Unsupported action %s", action))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	ctx = newContext(r, w, action)
+	defer logger.AuditLog(w, r, action, nil)
+
+	apReqBase64Str := r.Form.Get("APReq")
+	if apReqBase64Str == "" {
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSMissingParameter))
+		return
+	}
+
+	apReqBytes, err := base64.StdEncoding.DecodeString(apReqBase64Str)
+	if err != nil {
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	var APReq messages.APReq
+	err = APReq.Unmarshal(apReqBytes)
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("Error unmarshalling AP_REQ: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	ok, creds, err := service.VerifyAPREQ(APReq, globalServerConfig.Kerberos.settings)
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("Authentication failure: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+	if !ok {
+		logger.LogIf(ctx, fmt.Errorf("Authentication not accepted: %v", err))
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	////////////////////// TODO: DEBUG statements
+	fmt.Println("Group memberships in the credentials:", creds.AuthzAttributes())
+	fmt.Println("Attributes in the credentials:", creds.Attributes())
+	////////////////////// End Debug
+
+	if creds.Expired() {
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSKerberosExpiredToken))
+		return
+	}
+
+	m := map[string]interface{}{
+		"exp":           APReq.Ticket.DecryptedEncPart.EndTime,
+		"kerberosPrinc": fmt.Sprintf("%s@%s", strings.Join(creds.CName().NameString, "/"), creds.Realm()),
+
+		"policy": "readwrite",
```

Comment:
can you provide the status of this functionality?

Reply:
In this version, we just add read-write policy on the temporary credentials. Subsequently I shall add support for using a "default" policy if one is configured, as well as reading the group info from kerberos to pick a policy name equal to a group.